src/com/sap/piper/DefaultValueCache.groovy

```diff
@@ -18,7 +31,9 @@ class DefaultValueCache implements Serializable {
     }
 
     static createInstance(Map defaultValues){
-        instance = new DefaultValueCache(defaultValues)
+        instance = new DefaultValueCache(
+            immutable(defaultValues)
```

Comment:
Hm. I found that we are manipulating the default values at multiple places:
https://github.com/SAP/cloud-s4-sdk-pipeline-lib/blob/4a77b8bcb236c808bd80f78eade769e4336d7f2b/vars/setupDownloadCache.groovy#L17
https://github.com/SAP/cloud-s4-sdk-pipeline-lib/blob/4a77b8bcb236c808bd80f78eade769e4336d7f2b/vars/setupDownloadCache.groovy#L17

The reason is that we have default values that depend on the configuration of the Jenkins.

Reply:
Does the [second example](https://github.com/SAP/cloud-s4-sdk-pipeline-lib/blob/4a77b8bcb236c808bd80f78eade769e4336d7f2b/vars/setupDownloadCache.groovy#L23) above really affect the default values?

The funny thing is that the first example is annotated with a fixme comment (Here we misuse the defaultConfiguration ...)

The point is: default values are shared over several steps. Having the defaults mutable opens a huge area for introducing bugs. These bugs are not obvious and hard to troubleshoot, as we already saw with the phantom stash in neoDeploy. I think the normal mind set when developing is "Default Values? Well, they are stable ...". But with the current approach they can be manipulated by error. See our issue with stashing and neoDeploy ... Having our ConfigMerger approach it is fully intransparent that certain mutable collections etc. comes without defensive copy from the default values.

We can discuss if we would like to continue with mutable default values. My personal opinion is: we will have less trouble if they are immutable. This change is not provided for fun or for the sake of sticking to some academical coding principles.

@OliverNocon already suggested to provide a deep defensive copy of the configuration when calling `ConfigurationHelper#use()`. That copy can be - of course - mutable. With that we avoid updating default values from one step so that update is visible to other steps. That would be IMO also an improvement, also reducing the risk of unplanned config updates. But again: in my personal opinion our life is less complicated if the default values are immutable. I will provide another PR creating defensive copies in `ConfigurationHelper#use(./.)`

Reply:
I think as we can still create a new instance, I am fine with the change.